posthog/hogql/printer.py

```diff
@@ -1,196 +1,298 @@
-        if dialect == "clickhouse" and from_table is not None:
-            where = guard_where_team_id(where, context)
-        where = print_ast(where, stack, context, dialect) if where else None
-
-        having = print_ast(node.having, stack, context, dialect) if node.having else None
-        prewhere = print_ast(node.prewhere, stack, context, dialect) if node.prewhere else None
-        group_by = [print_ast(column, stack, context, dialect) for column in node.group_by] if node.group_by else None
-        order_by = [print_ast(column, stack, context, dialect) for column in node.order_by] if node.order_by else None
 
-        limit = node.limit
-        if context.limit_top_select:
-            if limit is not None:
-                if isinstance(limit, ast.Constant) and isinstance(limit.value, int):
-                    limit.value = min(limit.value, MAX_SELECT_RETURNED_ROWS)
+        select_from = []
+        next_join = node.select_from
+        while isinstance(next_join, ast.JoinExpr):
+            if next_join.symbol is None:
+                raise ValueError("Printing queries with a FROM clause is not permitted before symbol resolution")
+
+            visited_join = self.visit_join_expr(next_join)
+            select_from.append(visited_join.printed_sql)
+
+            # This is an expression we must add to the SELECT's WHERE clause to limit results.
+            extra_where = visited_join.where
+            if extra_where is None:
+                pass
+            elif isinstance(extra_where, ast.Expr):
+                if where is None:
+                    where = extra_where
+                elif isinstance(where, ast.And):
+                    where = ast.And(exprs=[extra_where] + where.exprs)
                 else:
-                    limit = ast.Call(name="min2", args=[ast.Constant(value=MAX_SELECT_RETURNED_ROWS), limit])
-            elif len(stack) == 1:
-                limit = ast.Constant(value=MAX_SELECT_RETURNED_ROWS)
+                    where = ast.And(exprs=[extra_where, where])
+            else:
+                raise ValueError(f"Invalid where of type {type(extra_where).__name__} returned by join_expr")
+
+            next_join = next_join.next_join
+
+        columns = [self.visit(column) for column in node.select] if node.select else ["1"]
+        where = self.visit(where) if where else None
+        having = self.visit(node.having) if node.having else None
+        prewhere = self.visit(node.prewhere) if node.prewhere else None
+        group_by = [self.visit(column) for column in node.group_by] if node.group_by else None
+        order_by = [self.visit(column) for column in node.order_by] if node.order_by else None
 
         clauses = [
             f"SELECT {'DISTINCT ' if node.distinct else ''}{', '.join(columns)}",
-            f"FROM {from_table}" if from_table else None,
+            f"FROM {' '.join(select_from)}" if len(select_from) > 0 else None,
```

Comment:
how does this work, is it valid syntax with a space in the middle? 🤔

Reply:
ah, it's the join clauses in here.

Name is a bit confusing, maybe `joined_tables` ? Actually, when there are no joins, how is the table name populated here..

Reply:
Renamed to `joined_tables`. The first table in a select is effectively a `ast.JoinExpr` without a `join_type` and `constraint` clause. Otherwise all the semantics are the same.